README.md
<div align="center">
  👩🏽‍💻🎒
</div>

<h1 align="center">
  devtogo
</h1>

<p align="center">
    a <a href="https://dev.to/">dev.to</a> tool for the road🧳
</p>

<div align="center">
  <a alt="GitHub Actions" href="https://github.com/softprops/devtogo/actions">
    <img src="https://github.com/softprops/devtogo/workflows/Main/badge.svg"/>
  </a>
  <a alt="crates.io" href="https://crates.io/crates/devtogo">
    <img src="https://img.shields.io/crates/v/devtogo.svg?logo=rust"/>
  </a>
  <a alt="license" href="LICENSE">
    <img src="https://img.shields.io/badge/license-MIT-brightgreen.svg"/>
  </a>
</div>

<br />

<img width="30px" src="https://cdn.svgporn.com/logos/markdown.svg"/> devtogo is a cli for publishing offline markdown files to dev.to. 

## 💾 install

### 🍻 [Homebrew](https://brew.sh/) (osx)

```sh
$ brew install softprops/tools/devtogo
```

### 📦 [Cargo](https://doc.rust-lang.org/cargo/)

```sh
$ cargo install devtogo
```

### 🚢 [GitHub Releases](https://help.github.com/github/administering-a-repository/about-releases)

Using a version from this repo's [GitHub Releases page](https://github.com/softprops/devtogo/releases),
substitute a VERSION below with a valid release version

```sh
$ cd $HOME/bin
$ export VERSION=v0.1.0
$ curl -L "https://github.com/softprops/devtogo/releases/download/${VERSION}/devtogo-$(uname -s)-$(uname -m).tar.gz" \
  | tar -xz -C ~/bin
```

## 🤸‍♂️ usage

devtogo is a cli for publishing offline markdown files to [dev.to](https://dev.to/). These files follow the same formatting rules documented in [dev.to's editor guide](https://dev.to/p/editor_guide).

At a bare minimum you'll want to declare a **`title`** in a [frontmatter section](https://jekyllrb.com/docs/front-matter/) of your markdown file.

```md
---
title: my very first post
---
# hello everybody
```


To get started, you will first need to export an **`DEVTO_API_KEY`** environment variable. You can generate yours 👉[here](https://dev.to/settings/account)

```sh
$ export DEVTO_API_KEY='som3R@ndOmAp1K3y'
```

The most basic usage is to run the program inside the directory containing your target article content

```sh
$ devtogo
```

This will scan the current working directory for articles: markdown documents containing frontmatter describing metadata about the article. devtogo uses the **`title`** frontmatter field as a unique identifier to compare existing remote content. 

When it can't resolve an article by title it uploads it, assuming its a new article. When it can, it local and remote compares content and uploads local copy if the content of the local copy differs.

> you can use the `published` frontmatter to indicate if and when an article should be published
  by default articles are saved as drafts only you can see. Setting published to true will publish articles.
  If you do this by accident you can set published back to false to unpublish an article if needed

To be more explicit you can provide a source argument which provides a path where content
is stored.

```sh
$ devtogo --source path/to/content
```

You can also experiment without actually posting your content using the `--dryrun` flag. This will perform all operations
except for uploading your content. This may be useful for validating your content.

```sh
$ devto --dryrun
```

You can always review the posts uploaded in [your dev.to dashboard online](https://dev.to/dashboard)

Doug Tangren (softprops) 2020


CHANGELOG.md
# 0.1.0

* initial release

# 0.0.2

* filter files in source directory to those ending with markdown extension [#2](https://github.com/softprops/devtogo/pull/2)

# 0.0.1

* test release

src/new.rs
use dialoguer::{theme::ColorfulTheme, Input};
use structopt::StructOpt;

/// 🆕 Creates a new local article md file to get you going
#[derive(StructOpt, Debug)]
pub struct New {
    /// title for article
    #[structopt(long)]
    title: Option<String>,
}

pub async fn run(new: New) -> anyhow::Result<()> {
    let New { title, .. } = new;
    let theme = ColorfulTheme::default();
    let title: String = title
        .ok_or_else(|| anyhow::anyhow!("article title is required"))
        .or_else::<anyhow::Error, _>(|_| {
            Ok(Input::<String>::with_theme(&theme)
                .with_prompt("article title")
                .interact()?)
        })?;
    println!(
        r#"---
title: {title}
published: false
# tags: foo, bar
# date: ...
# series: ...
# canonical_url: ..
# cover_image: ...
---
"#,
        title = title
    );
    Ok(())
}


src/sync.rs
use chrono::DateTime;
use colored::Colorize;
use frontmatter::Yaml;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{fmt, fs, path::PathBuf};
use structopt::StructOpt;
use walkdir::WalkDir;

enum Status<'a> {
    Uploaded,
    Syncing(&'a Article),
    Posting,
}

impl fmt::Display for Status<'_> {
    fn fmt(
        &self,
        f: &mut fmt::Formatter<'_>,
    ) -> fmt::Result {
        let s = format!(
            "{}",
            match self {
                Status::Uploaded => "UPLOADED".green(),
                Status::Posting => "POSTING".yellow(),
                Status::Syncing(ref _remote) => "SYNCING".yellow(),
            }
        );
        f.write_str(&s)
    }
}

#[derive(Debug, Deserialize)]
struct DevtoError {
    error: String,
}

#[derive(Debug, Serialize)]
struct CreateArticleInput {
    body_markdown: String,
}

#[derive(Debug, Deserialize)]
struct Article {
    id: u32,
    title: String,
    description: String,
    cover_image: Option<String>,
    published: bool,
    published_at: Option<String>,
    tag_list: Vec<String>,
    slug: String,
    path: String,
    url: String,
    canonical_url: String,
    published_timestamp: String,
    body_markdown: String,
}

/// 🔼 Uploads local markdown files with dev.to
#[derive(StructOpt, Debug)]
pub struct Sync {
    /// Directory to source markdown files from. Defaults to current working directory
    #[structopt(short, long)]
    source: Option<PathBuf>,
    /// Run without actually updating account
    #[structopt(short, long)]
    dryrun: bool,
}

fn extract(
    name: &str,
    content: &str,
) -> anyhow::Result<(Frontmatter, String)> {
    let (front, back) = frontmatter::parse_and_find_content(&content)?;
    let metadata = front.ok_or_else(
        || {
            anyhow::anyhow!(
                "file {} is missing required markdown frontmatter.\n  ▶ Please see https://dev.to/p/editor_guide more information on what frontmatter is expected", name
            )
        }
    )?;

    Ok((Frontmatter::from_file(&name, metadata)?, back.into()))
}

/// Markdown frontmatter dev.to api documents as acceptable input
#[derive(Debug, PartialEq, Default)]
struct Frontmatter {
    title: String,
    published: Option<bool>,
    tags: Option<String>,
    date: Option<String>,
    series: Option<String>,
    canonical_url: Option<String>,
    cover_image: Option<String>,
}

impl Frontmatter {
    /// extract and validate raw yaml frontmatter
    fn from_file(
        name: &str,
        metadata: Yaml,
    ) -> anyhow::Result<Frontmatter> {
        let hash = metadata.into_hash().ok_or_else(|| {
            anyhow::anyhow!("file {} contains frontmatter that not well formatted", name)
        })?;
        let string = |name: &str| -> Option<String> {
            hash.get(&Yaml::String(name.into()))
                .and_then(|v| v.as_str().map(|s| s.into()))
        };
        let boolean = |name: &str| -> Option<bool> {
            hash.get(&Yaml::String(name.into()))
                .and_then(|v| v.as_bool())
        };
        let title = string("title").ok_or_else(|| {
            anyhow::anyhow!("file {} contains frontmatter missing a string title", name)
        })?;
        let published = boolean("published");
        let tags = string("tags");
        let date = string("date");
        if let Some(value) = &date {
            if DateTime::parse_from_rfc3339(&value).is_err() {
                anyhow::bail!(
                    "file {} contains frontmatter with and invalid date: {}",
                    name,
                    value
                );
            }
        }
        let series = string("series");
        let canonical_url = string("canonical_url");
        let cover_image = string("cover_image");

        Ok(Frontmatter {
            title,
            published,
            tags,
            date,
            series,
            canonical_url,
            cover_image,
        })
    }
}

async fn post(
    client: Client,
    api_key: String,
    content: String,
) -> anyhow::Result<()> {
    again::retry(move || {
        let client = client.clone();
        let api_key = api_key.clone();
        let content = content.clone();
        async move {
            let resp = client
                .post("https://dev.to/api/articles")
                .header("api-key", api_key.as_str())
                .json(&CreateArticleInput {
                    body_markdown: content,
                })
                .send()
                .await?;

            if !resp.status().is_success() {
                println!("Dev.to error: {:#?} {}", resp.status(), resp.text().await?);
            } else {
                println!("Post was successful");
            }
            Ok(())
        }
    })
    .await
}

async fn put(
    id: u32,
    client: Client,
    api_key: String,
    content: String,
) -> anyhow::Result<()> {
    again::retry(move || {
        let client = client.clone();
        let api_key = api_key.clone();
        let content = content.clone();
        async move {
            let resp = client
                .put(format!("https://dev.to/api/articles/{}", id).as_str())
                .header("api-key", api_key.as_str())
                .json(&CreateArticleInput {
                    body_markdown: content,
                })
                .send()
                .await?;

            if !resp.status().is_success() {
                println!("Dev.to error {:#?} {}", resp.status(), resp.text().await?);
            } else {
                println!("Update was successful");
            }
            Ok(())
        }
    })
    .await
}

async fn fetch(
    client: &Client,
    api_key: &str,
) -> anyhow::Result<Vec<Article>> {
    Ok(client
        .get("https://dev.to/api/articles/me/all?per_page=1000")
        .header("api-key", api_key)
        .send()
        .await?
        .json()
        .await?)
}

pub async fn run(
    api_key: String,
    args: Sync,
) -> anyhow::Result<()> {
    let Sync { source, dryrun } = args;
    let client = Client::new();
    let articles = fetch(&client, &api_key).await?;
    let mut hasher = Sha256::new();
    for path in WalkDir::new(source.unwrap_or_else(|| ".".into()))
        .into_iter()
        .filter_map(|e| e.ok().map(|e| e.path().to_owned()).filter(|p| !p.is_dir()))
    {
        let client = client.clone();
        let api_key = api_key.clone();
        let content = fs::read_to_string(&path)?;
        let name = path
            .as_path()
            .file_name()
            .unwrap_or_default()
            .to_string_lossy();
        let (meta, _) = extract(name.as_ref(), &content)?;
        let status = match articles.iter().find(|a| a.title == meta.title) {
            None => Status::Posting,
            Some(remote) => {
                let differ = {
                    hasher.update(content.as_bytes());
                    let local = hasher.finalize_reset();
                    hasher.update(remote.body_markdown.as_bytes());
                    let remote = hasher.finalize_reset();
                    local != remote
                };
                if differ {
                    Status::Syncing(remote)
                } else {
                    Status::Uploaded
                }
            }
        };
        println!(
            "{}{}{}",
            meta.title.chars().take(50).collect::<String>().bold(),
            String::from(".").repeat(50 - meta.title.len()).dimmed(),
            format!("[{}]", status).bold()
        );
        if !dryrun {
            match status {
                Status::Syncing(remote) => {
                    put(remote.id, client.clone(), api_key.clone(), content.clone()).await?
                }
                Status::Posting => post(client.clone(), api_key.clone(), content.clone()).await?,
                _ => (),
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn status_impl_display() {
        fn test(_: impl fmt::Display) {}
        test(Status::Posting)
    }

    #[test]
    fn test_extract_fails_with_missing_frontmatter() {
        let result = extract(
            "foo.md",
            r#"
        "#,
        );
        assert!(result.is_err())
    }

    #[test]
    fn test_extract_fails_with_missing_title() {
        let result = extract(
            "foo.md",
            r#"
        --
        --
        "#,
        );
        assert!(result.is_err())
    }

    #[test]
    fn test_extract_passes_with_missing_frontmatter() -> anyhow::Result<()> {
        let (front, _) = extract(
            "foo.md",
            r#"---
            title: foo
            ---
            "#,
        )?;
        assert_eq!(
            front,
            Frontmatter {
                title: "foo".into(),
                ..Frontmatter::default()
            }
        );
        Ok(())
    }

    #[test]
    fn test_extract_validates_date() {
        let result = extract(
            "foo.md",
            r#"---
            title: foo
            date: ...
            ---
            "#,
        );
        assert!(result.is_err());
    }
}


src/push.rs
use anyhow::{anyhow, bail};
use chrono::DateTime;
use colored::Colorize;
use frontmatter::Yaml;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{fmt, fs, path::PathBuf};
use structopt::StructOpt;
use walkdir::WalkDir;

enum UploadStatus<'a> {
    Uploaded,
    Syncing(&'a Article),
    Posting,
}

impl fmt::Display for UploadStatus<'_> {
    fn fmt(
        &self,
        f: &mut fmt::Formatter<'_>,
    ) -> fmt::Result {
        let s = format!(
            "{}",
            match self {
                UploadStatus::Uploaded => "UPLOADED".green(),
                UploadStatus::Posting => "POSTING".yellow(),
                UploadStatus::Syncing(ref _remote) => "SYNCING".yellow(),
            }
        );
        f.write_str(&s)
    }
}

#[derive(PartialEq, Debug)]
enum PublishStatus {
    Published,
    Draft,
}

impl fmt::Display for PublishStatus {
    fn fmt(
        &self,
        f: &mut fmt::Formatter<'_>,
    ) -> fmt::Result {
        let s = format!(
            "{}",
            match self {
                PublishStatus::Published => "published".dimmed(),
                PublishStatus::Draft => "draft".dimmed(),
            }
        );
        f.write_str(&s)
    }
}

#[derive(Debug, Deserialize)]
struct DevtoError {
    error: String,
}

#[derive(Debug, Serialize)]
struct CreateArticleInput {
    body_markdown: String,
}

#[derive(Debug, Deserialize)]
struct Article {
    id: u32,
    title: String,
    description: String,
    cover_image: Option<String>,
    published: bool,
    published_at: Option<String>,
    tag_list: Vec<String>,
    slug: String,
    path: String,
    url: String,
    canonical_url: String,
    published_timestamp: String,
    body_markdown: String,
}

/// A dev.to tool for the road 👩🏽‍💻🎒
///
/// Uploads local markdown files with dev.to
#[derive(StructOpt, Debug)]
pub struct Push {
    /// Directory to source markdown files from. Defaults to current working directory
    #[structopt(short, long)]
    source: Option<PathBuf>,
    /// Run without actually updating account
    #[structopt(short, long)]
    dryrun: bool,
}

fn extract(
    name: &str,
    content: &str,
) -> anyhow::Result<(Frontmatter, String)> {
    let (front, back) = frontmatter::parse_and_find_content(&content)?;
    let metadata = front.ok_or_else(
        || {
            anyhow!(
                "file {} is missing required markdown frontmatter.\n  ▶ Please see https://dev.to/p/editor_guide more information on what frontmatter is expected", name
            )
        }
    )?;

    Ok((Frontmatter::from_file(&name, metadata)?, back.into()))
}

/// Markdown frontmatter dev.to api documents as acceptable input
#[derive(Debug, PartialEq, Default)]
struct Frontmatter {
    title: String,
    published: Option<bool>,
    tags: Option<String>,
    date: Option<String>,
    series: Option<String>,
    canonical_url: Option<String>,
    cover_image: Option<String>,
}

impl Frontmatter {
    fn publish_status(&self) -> PublishStatus {
        if self.published.unwrap_or_default() {
            PublishStatus::Published
        } else {
            PublishStatus::Draft
        }
    }
    /// extract and validate raw yaml frontmatter
    fn from_file(
        name: &str,
        metadata: Yaml,
    ) -> anyhow::Result<Frontmatter> {
        let hash = metadata
            .into_hash()
            .ok_or_else(|| anyhow!("file {} contains frontmatter that not well formatted", name))?;
        let string = |name: &str| -> Option<String> {
            hash.get(&Yaml::String(name.into()))
                .and_then(|v| v.as_str().map(|s| s.into()))
        };
        let boolean = |name: &str| -> Option<bool> {
            hash.get(&Yaml::String(name.into()))
                .and_then(|v| v.as_bool())
        };
        let title = string("title")
            .ok_or_else(|| anyhow!("file {} contains frontmatter missing a string title", name))?;
        let published = boolean("published");
        let tags = string("tags");
        let date = string("date");
        if let Some(value) = &date {
            if DateTime::parse_from_rfc3339(&value).is_err() {
                bail!(
                    "file {} contains frontmatter with and invalid date: {}",
                    name,
                    value
                );
            }
        }
        let series = string("series");
        let canonical_url = string("canonical_url");
        let cover_image = string("cover_image");

        Ok(Frontmatter {
            title,
            published,
            tags,
            date,
            series,
            canonical_url,
            cover_image,
        })
    }
}

async fn post(
    client: Client,
    api_key: String,
    content: String,
) -> anyhow::Result<()> {
    again::retry(move || {
        let client = client.clone();
        let api_key = api_key.clone();
        let content = content.clone();
        async move {
            let resp = client
                .post("https://dev.to/api/articles")
                .header("api-key", api_key.as_str())
                .json(&CreateArticleInput {
                    body_markdown: content,
                })
                .send()
                .await?;

            if !resp.status().is_success() {
                println!("Dev.to error: {:#?} {}", resp.status(), resp.text().await?);
            } else {
                println!("Post was successful");
            }
            Ok(())
        }
    })
    .await
}

async fn put(
    id: u32,
    client: Client,
    api_key: String,
    content: String,
) -> anyhow::Result<()> {
    again::retry(move || {
        let client = client.clone();
        let api_key = api_key.clone();
        let content = content.clone();
        async move {
            let resp = client
                .put(format!("https://dev.to/api/articles/{}", id).as_str())
                .header("api-key", api_key.as_str())
                .json(&CreateArticleInput {
                    body_markdown: content,
                })
                .send()
                .await?;

            if !resp.status().is_success() {
                println!("Dev.to error {:#?} {}", resp.status(), resp.text().await?);
            } else {
                println!("Update was successful");
            }
            Ok(())
        }
    })
    .await
}

async fn fetch(
    client: &Client,
    api_key: &str,
) -> anyhow::Result<Vec<Article>> {
    Ok(client
        .get("https://dev.to/api/articles/me/all?per_page=1000")
        .header("api-key", api_key)
        .send()
        .await?
        .json()
        .await?)
}

fn valid_path(path: &PathBuf) -> bool {
    !path.is_dir()
        && path
            .extension()
            .into_iter()
            .any(|e| e == "md" || e == "markdown")
}

pub async fn run(
    api_key: String,
    args: Push,
) -> anyhow::Result<()> {
    let Push { source, dryrun } = args;
    let client = Client::new();
    let articles = fetch(&client, &api_key).await?;
    let mut hasher = Sha256::new();
    for path in WalkDir::new(source.unwrap_or_else(|| ".".into()))
        .into_iter()
        .filter_map(|e| e.ok().map(|e| e.path().to_path_buf()))
        .filter(valid_path)
    {
        let client = client.clone();
        let api_key = api_key.clone();
        let content = fs::read_to_string(&path)?;
        let name = path.file_name().unwrap_or_default().to_string_lossy();
        let (meta, _) = extract(name.as_ref(), &content)?;
        let status = match articles.iter().find(|a| a.title == meta.title) {
            None => UploadStatus::Posting,
            Some(remote) => {
                let differ = {
                    hasher.update(content.as_bytes());
                    let local = hasher.finalize_reset();
                    hasher.update(remote.body_markdown.as_bytes());
                    let remote = hasher.finalize_reset();
                    local != remote
                };
                if differ {
                    UploadStatus::Syncing(remote)
                } else {
                    UploadStatus::Uploaded
                }
            }
        };
        println!(
            "{}{}{}",
            meta.title.chars().take(50).collect::<String>().bold(),
            String::from(".")
                .repeat(50_usize.checked_sub(meta.title.len()).unwrap_or_default())
                .dimmed(),
            format!("[{} {}]", status, meta.publish_status()).bold(),
        );
        if !dryrun {
            match status {
                UploadStatus::Syncing(remote) => {
                    put(remote.id, client.clone(), api_key.clone(), content.clone()).await?
                }
                UploadStatus::Posting => {
                    post(client.clone(), api_key.clone(), content.clone()).await?
                }
                _ => (),
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn frontmatter_is_published_when_expected() {
        assert_eq!(
            Frontmatter::default().publish_status(),
            PublishStatus::Draft
        );
        assert_eq!(
            Frontmatter {
                published: Some(false),
                ..Frontmatter::default()
            }
            .publish_status(),
            PublishStatus::Draft
        );
        assert_eq!(
            Frontmatter {
                published: Some(true),
                ..Frontmatter::default()
            }
            .publish_status(),
            PublishStatus::Published
        );
    }
    #[test]
    fn valid_path_isnt_dirs() {
        assert!(!valid_path(&PathBuf::from("/")))
    }

    #[test]
    fn valid_path_contains_md_ext() {
        assert!(valid_path(&PathBuf::from("/foo.md")));
    }

    #[test]
    fn valid_path_contains_markdown_ext() {
        assert!(valid_path(&PathBuf::from("/foo.markdown")));
    }

    #[test]
    fn valid_path_doesnt_contains_other_ext() {
        assert!(!valid_path(&PathBuf::from("/foo.txt")));
    }

    #[test]
    fn upload_status_impl_display() {
        fn test(_: impl fmt::Display) {}
        test(UploadStatus::Posting)
    }

    #[test]
    fn publish_status_impl_display() {
        fn test(_: impl fmt::Display) {}
        test(PublishStatus::Draft)
    }

    #[test]
    fn test_extract_fails_with_missing_frontmatter() {
        let result = extract(
            "foo.md",
            r#"
        "#,
        );
        assert!(result.is_err())
    }

    #[test]
    fn test_extract_fails_with_missing_title() {
        let result = extract(
            "foo.md",
            r#"
        --
        --
        "#,
        );
        assert!(result.is_err())
    }

    #[test]
    fn test_extract_passes_with_missing_frontmatter() -> anyhow::Result<()> {
        let (front, _) = extract(
            "foo.md",
            r#"---
            title: foo
            ---
            "#,
        )?;
        assert_eq!(
            front,
            Frontmatter {
                title: "foo".into(),
                ..Frontmatter::default()
            }
        );
        Ok(())
    }

    #[test]
    fn test_extract_validates_date() {
        let result = extract(
            "foo.md",
            r#"---
            title: foo
            date: ...
            ---
            "#,
        );
        assert!(result.is_err());
    }
}
